conformance/tests/generics_self_advanced.py

```diff
@@ -0,0 +1,46 @@
+"""
+Tests for advanced or special-case usage of the typing.Self type.
+"""
+
+from typing import assert_type, Self
+
+
+class ParentA:
+    # Test for property that returns Self.
+    @property
+    def prop1(self) -> Self:
+        ...
+
+class ChildA(ParentA):
+    ...
+
+
+assert_type(ParentA().prop1, ParentA)
+assert_type(ChildA().prop1, ChildA)
+
+
+# Test for a child that accesses an attribute within a parent
+# whose type is annotated using Self.
+class ParentB:
+    a: list[Self]
+
+    @classmethod
+    def method1(cls) -> Self:
+        ...
+
+class ChildB(ParentB):
+    b: int = 0
+
+    def method2(self) -> None:
+        assert_type(self, Self)
```

Comment:
Is this actually specified? The spec says:

> One implementation strategy is to simply desugar the former to the latter in a preprocessing step. If a method uses Self in its signature, the type of self within a method will be Self. In other cases, the type of self will remain the enclosing class.

Same question for the classmethod variation

Reply:
The spec may not say this explicitly (it probably should), but I don't know how you could correctly implement the `Self` type if an unannotated `self` doesn't have this type.

Consider this example, which I think should type check without errors given the spec.

```python
class Foo:
    def method1(self) -> Self:
        reveal_type(self)
        return self

    def method2(self) -> Any:
        reveal_type(self)
        x: Self = self
        return x
```

Reply:
If your reading of the spec is different, then let's ignore (or comment out) this part of the conformance test for now and queue the topic for discussion.

Reply:
Your example should type check and I'm fine with specifying this behaviour. But yes, my reading of the current spec is different. In general, I think the construction of the conformance suite should be used as a way to iron out ambiguity in the spec.

https://github.com/python/mypy/issues/14075 is the relevant mypy issue